src/go/k8s/pkg/resources/statefulset_update.go

```diff
@@ -0,0 +1,253 @@
+			return err
+		}
+
+		r.logger.Info("Waiting for pod to complete updating", "partition:", ordinal)
+
+		if err := r.waitUntilPodImageIsVerified(ctx, sts, newImage, ordinal); err != nil {
+			return err
+		}
+
+		// Give time to redpanda pod to rejoin cluster.
+		// In a revision, an API call to redpanda is made to help us decide when to continue.
+		r.logger.Info("Sleeping for 1 minute before updating the next pod, or completion")
+		time.Sleep(time.Minute)
+
+		// After Update(), get the statefulset for next iteration (or other use).
+		err := r.Get(ctx, r.Key(), sts)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) podImageIdenticalToClusterImage(
+	ctx context.Context, sts *appsv1.StatefulSet, newImage string, ordinal int32,
+) error {
+	podName := fmt.Sprintf("%s-%d", sts.Name, ordinal)
+
+	var pod corev1.Pod
+	if err := r.Get(ctx, types.NamespacedName{Name: podName, Namespace: sts.Namespace}, &pod); err != nil {
+		return err
+	}
+
+	container, err := findContainer(pod.Spec.Containers, redpandaContainerName)
+	if err != nil {
+		return err
+	}
+
+	if container.Image != newImage {
+		r.logger.Info("Container image not updated to cluster image", "pod", pod.Name,
+			"container", container.Name, "container image", container.Image, "cluster image", newImage)
+		return containerHasWrongImageError(podName, container.Name, container.Image, newImage)
+	}
+
+	container, err = findContainer(pod.Spec.InitContainers, configuratorContainerName)
+	if err != nil {
+		return err
+	}
+
+	if container.Image != newImage {
+		r.logger.Info("Init container image not updated to cluster image", "pod", pod.Name,
+			"container", container.Name, "container image", container.Image, "cluster image", newImage)
+		return containerHasWrongImageError(podName, container.Name, container.Image, newImage)
+	}
+
+	if !podIsReady(&pod) {
+		r.logger.Info("Pod not ready yet", "pod", pod.Name)
+		return podNotReadyError(pod.Name)
+	}
+
+	r.logger.Info(fmt.Sprintf("Pod (%s) is ready, with cluster image (%s)", pod.Name, container.Image))
+
+	return nil
+}
+
+func (r *StatefulSetResource) rollingUpdatePartition(
+	ctx context.Context, sts *appsv1.StatefulSet, ordinal int32,
+) error {
+	r.logger.Info("Call update on statefulset", "partition", ordinal)
+
+	sts.Spec.UpdateStrategy.RollingUpdate = &appsv1.RollingUpdateStatefulSetStrategy{
+		Partition: &ordinal,
+	}
+	if err := r.Update(ctx, sts); err != nil {
+		return fmt.Errorf("failed to update StatefulSet (partition %d): %w", ordinal, err)
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) waitUntilPodImageIsVerified(
+	ctx context.Context, sts *appsv1.StatefulSet, image string, ordinal int32,
+) error {
+	f := func() (bool, error) {
+		err := r.podImageIdenticalToClusterImage(ctx, sts, image, ordinal)
+		if err != nil {
+			// Retry
+			return false, nil
+		}
+
+		return true, nil
+	}
+
+	return wait.ExponentialBackoff(backoff, f)
```

Comment:
Should this be truncated? i.e., limited to 5 mins or something. When I've operated services in the past, I've always been tempted to kick pods manually as this can get really high, really quick.

Reply:
It currently has a `waitCap` of 5 minutes. It's adjustable.